dieboljo/embedded-loop-station
# Feature/unit tests

## Related Issues:
* This is unrelated to a specific ticket, and instead updates the project infrastructure.

## Main Changes:
* Migrated the project structure to that specified in the [PlatformIO](https://docs.platformio.org/en/latest/) framework.
* Rolled back the main application source code to a functional version, and started implementing the previously attempted classes into separate, testable libraries.
* Added a basic playback unit test, to demonstrate the unit test structure.

## Steps To Test:
1. Head to **PlatformIO**'s installation instructions, and install the necessary VS Code extension (that should be all that is necessary).
2. With this branch checked out, open the _app_ folder (under the root directory) in VS Code.
3. The extension should recognize the project as a **PlatformIO** project.
4. The status bar at the bottom has quick access to necessary project functionality:
![screenshot](https://user-images.githubusercontent.com/71578303/218612468-c058fbaf-f304-451b-b1fc-a709f971a3a9.png)
5. Click the check mark to build the main source code.
6. If the necessary hardware is available, click the right arrow (->) to upload the compiled project to the device.
7. Click the flask icon to run unit tests (all, by default). Currently, these tests can only be run on an attached Teensy device.

## TODO
- [ ] I retained some example files to aid further use of the framework. These should be removed in the future.
- [ ] I also included the `.ccls` commands, which aid the language server I use for **Vim**. These probably don't need to be included, as they aren't need for the IDE, and they can be removed in the future as well.
